src/sc/sidechainTxsCommitmentGuard.h

```diff
@@ -0,0 +1,90 @@
+#ifndef SIDECHAIN_TX_COMMITMENT_GUARD
+#define SIDECHAIN_TX_COMMITMENT_GUARD
+
+#include "coins.h"
+#include <sc/sidechaintypes.h>
+
+class uint256;
+class CTransaction;
+
+class CScCertificate;
+class CTxForwardTransferOut;
+class CBwtRequestOut;
+class CTxCeasedSidechainWithdrawalInput;
+
+class SidechainTxsCommitmentGuard
+{
+public:
+    SidechainTxsCommitmentGuard() {};
+    ~SidechainTxsCommitmentGuard() {};
+
+    SidechainTxsCommitmentGuard(const SidechainTxsCommitmentGuard&) = delete;
+    SidechainTxsCommitmentGuard& operator=(const SidechainTxsCommitmentGuard&) = delete;
+
+    bool add(const CTransaction& tx, bool autoRewind = false);
+    bool add(const CScCertificate& cert);
+
+private:
+    bool add_fwt(const CTxForwardTransferOut& ccout);
+    bool add_bwtr(const CBwtRequestOut& ccout);
+    bool add_csw(const CTxCeasedSidechainWithdrawalInput& ccin);
+    bool add_cert(const CScCertificate& cert);
+
+    void rewind(const CTransaction& tx, const int failingFt, const int failingBwtr, const int failingCsw);
+
+    // Keeping information separated to closely mimic CCTPlib structures
+    struct CommitmentBuilderStatsAliveCounter {
+        uint32_t   ft = 0;
+        uint32_t bwtr = 0;
+        uint32_t cert = 0;
+        uint32_t  bwt = 0;
+    };
+    struct CommitmentBuilderStatsCeasedCounter {
+        uint32_t  csw = 0;
+    };
+    struct CommitmentBuilderStats {
+        std::map<uint256, CommitmentBuilderStatsAliveCounter>  cbsaMap;
+        std::map<uint256, CommitmentBuilderStatsCeasedCounter> cbscMap;
+        // The following values MUST be aligned with those specified in CCTPlib!
+        static const int   SC_LIMIT = 4096;
+        static const int   FT_LIMIT = 4095;
+        static const int BWTR_LIMIT = 4095;
+        static const int CERT_LIMIT = 4095;
+        static const int  CSW_LIMIT = 4095;
+        static const int  BWT_LIMIT = 4096;
```

Comment:
I think we need to keep BWT_LIMIT = 4095 as well because of the way how it is handled by CCTP lib (limit errors are ignored there)
Beside that it seems the limit is considered as a total number of BWT per sidechain, but must be per certificate.
For example, cert A (sc1, bt_number = 3000), cert B(sc1, bt_number 2000) are allowed to be put inside the same block. It is not an issue for CommTree.

Reply:
Yes, you are right. I changed the logic so that we check the limit per certificate. Also, limit has been decreased as suggested.